connectors/spark/src/test/java/io/unitycatalog/spark/auth/AwsVendedTokenProviderTest.java

```diff
@@ -0,0 +1,329 @@
+        .hasMessage("'%s' is not set in hadoop configuration", UCHadoopConf.UC_TOKEN);
+  }
+
+  @Test
+  public void testTableTemporaryCredentialsRenew() throws Exception {
+    Configuration conf = new Configuration();
+    conf.set(UCHadoopConf.UC_URI, "http://localhost:8080");
+    conf.set(UCHadoopConf.UC_TOKEN, "unity-catalog-token");
+
+    // For table-based temporary requests.
+    conf.set(UCHadoopConf.UC_CREDENTIALS_TYPE, UCHadoopConf.UC_CREDENTIALS_TYPE_TABLE_VALUE);
+    conf.set(UCHadoopConf.UC_TABLE_ID, "test");
+    conf.set(UCHadoopConf.UC_TABLE_OPERATION, TableOperation.READ.getValue());
+
+    long expirationTime1 = System.currentTimeMillis() + 1000 + 3 * 1000L;
+    long expirationTime2 = System.currentTimeMillis() + 1000 + 6 * 1000L;
+    TemporaryCredentials cred1 =
+        newAwsTempCredentials("accessKeyId1", "secretAccessKey1", "sessionToken1", expirationTime1);
+    TemporaryCredentials cred2 =
+        newAwsTempCredentials("accessKeyId2", "secretAccessKey2", "sessionToken2", expirationTime2);
+
+    // Mock the table-based temporary credentials' generation.
+    TemporaryCredentialsApi tempCredApi = mock(TemporaryCredentialsApi.class);
+    when(tempCredApi.generateTemporaryTableCredentials(any())).thenReturn(cred1).thenReturn(cred2);
+
+    AwsVendedTokenProvider provider = new TestAwsVendedTokenProvider(null, conf, tempCredApi);
+    provider.setRenewalLeadTime(3 * 1000L);
+
+    // Use the cred1 for the 1st access.
+    AwsSessionCredentials awsCred1 = (AwsSessionCredentials) provider.resolveCredentials();
+    assertThat(awsCred1.accessKeyId()).isEqualTo("accessKeyId1");
+    assertThat(awsCred1.secretAccessKey()).isEqualTo("secretAccessKey1");
+    assertThat(awsCred1.sessionToken()).isEqualTo("sessionToken1");
+
+    // Use the cred1 for the 2nd access, since it's valid.
+    AwsSessionCredentials awsCred2 = (AwsSessionCredentials) provider.resolveCredentials();
+    assertThat(awsCred2.accessKeyId()).isEqualTo("accessKeyId1");
+    assertThat(awsCred2.secretAccessKey()).isEqualTo("secretAccessKey1");
+    assertThat(awsCred2.sessionToken()).isEqualTo("sessionToken1");
+
+    // Sleep 4 seconds to wait the cred1 to be expired.
+    Thread.sleep(4 * 1000L);
+
+    // Use the cred2 for the 3rd access, since cred1 it's expired.
+    AwsSessionCredentials awsCred3 = (AwsSessionCredentials) provider.resolveCredentials();
+    assertThat(awsCred3.accessKeyId()).isEqualTo("accessKeyId2");
+    assertThat(awsCred3.secretAccessKey()).isEqualTo("secretAccessKey2");
+    assertThat(awsCred3.sessionToken()).isEqualTo("sessionToken2");
+
+    // Use the cred3 for the 4th access, since cred2 is valid.
+    AwsSessionCredentials awsCred4 = (AwsSessionCredentials) provider.resolveCredentials();
+    assertThat(awsCred4.accessKeyId()).isEqualTo("accessKeyId2");
+    assertThat(awsCred4.secretAccessKey()).isEqualTo("secretAccessKey2");
+    assertThat(awsCred4.sessionToken()).isEqualTo("sessionToken2");
```

Comment:
i counted something like 16 sets of these 3-line checks = 48 lines. if you define a function, that can be reduced to 16 lines + 4 lines for a function.

please keep an eye out for reduce duplicate and verbose code.